skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Stores the position and metadata of an interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, interval_metadata=None):
+        iv = []
+        for interval in intervals:
+            iv.append(_polish_interval(interval))
+
+        if boundaries is not None:
+            self.boundaries = boundaries
+        else:
+            self.boundaries = []
+
+        if metadata is not None:
+            self.metadata = metadata
+        else:
+            self.metadata = {}
+
+        self._interval_metadata = interval_metadata
+
+        if intervals is not None:
+            self.intervals = sorted(iv)
+        else:
+            self.intervals = []
+
+        if interval_metadata is not None:
+            self._add_interval()
+
+    def _add_interval(self):
+        """ Add an interval object to the IntervalTree within
+            IntervalMetadata."""
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in self.intervals:
+            if loc is not None:
+                start, end = loc
+                self._interval_metadata._intervals.add(start, end, self)
+        self._interval_metadata._metadata.append(self)
+
+    def __getitem__(self, key):
```

Comment:
Since we don't know what `__iter__` and `__len__` are yet, I think `__getitem__` and `__setitem__` should be removed since the [container protocol](https://docs.python.org/3/reference/datamodel.html#emulating-container-types) methods need to be consistent. For now, users can get/set on `Interval.metadata`, which is a little more work but allows us to define the container magic methods consistently in the future without breaking backward-compatibility.